Adjust toHaveStyleRule method to use proper base component name (chan…

dumbNickname opened the pull request:
…ged in commit: https://github.com/styled-components/styled-components/commit/1c19edc77e9f4e1f4059dba4354b4984c02dd98f#diff-8a1d0988ee5ae37265b4647fa41def2a), disabled calling lifecycle methods for enzyme shallow rendering  - unfortunately this would need to be fixed in tests that use the library (no better solution yet).

dumbNickname commented:
Hey. Just created this commit to get quick feedback if changes are going in right direction. The most crutial change is rename of base component for styled-components v.4. For now I just changed to new one, but if backward compatibility is required then probably both cases should be considered. 

Another thing is necessity to provide ``{ disableLifecycleMethods: true }`` for shallow rendering (it will be a responsibility of users to update their tests - I do not know if this is acceptable).

There is still one test failing considering `` {"modifier":"& &"}`` but this is still in analysis -as I said I would appreciate feedback faster ;)

dumbNickname commented:
@MicheleBertoli Just made the next commit that I mentioned in the issue. This is still work in progress but makes more tests green. Btw. I do not know why travis build fails, but I assume it has something to do with some additional preact tests, which I did not yer reach :P

dumbNickname commented:
@MicheleBertoli one more question. I just enabled the test case for styled component nested in a wrapper. It fails for shallow rendering, but imho that seems to be correct as shallow does not resolve below the wrapper. How was it working before? Are there any other intentions behind this test case? Am I missing something?

MicheleBertoli commented on src/toHaveStyleRule.js:
```diff
@@ -54,28 +54,49 @@ const getAtRules = (ast, options) => {
-  }
-
-  return `${prefix}${modifier}`.trim()
+const getModifiedClassName = ({ className, modifier = '', componentId }) => {
```
I would revert this and remove the `& &` test case.

Previously, `& &` was rendered as:

```css
.lnlmMg .lnlmMg {
  color: deepskyblue;
}
```

While now it's rendered as:

```css
.icfrky .sc-jTzLTM {
  color: deepskyblue;
}
```

I wonder if this is intentional and what are the motivations behind this decision.
We should open an issue on the main repo.

MicheleBertoli commented:
Thank you very much for working on this, @dumbNickname - I really appreciate that.

> For now I just changed to new one, but if backward compatibility is required then probably both cases should be considered.

I think it's fine as `BaseStyledComponent` was used in the beta and we don't need to support that.

> I do not know why travis build fails, but I assume it has something to do with some additional preact tests

I can confirm the build is failing because I "disabled" the Preact tests - let's get the React ones green first.

> I just enabled the test case for styled component nested in a wrapper. It fails for shallow rendering, but imho that seems to be correct as shallow does not resolve below the wrapper. How was it working before? Are there any other intentions behind this test case? Am I missing something?

It works with v3 thanks to the way we are diving into the tree to find the class name.
I'm happy to avoid covering the case (at least for now).


I left a couple of comments in the code, please let me know if they are clear.
Cheers!

dumbNickname commented on test/toHaveStyleRule.spec.js:
```diff
@@ -196,7 +202,7 @@ test('styled child', () => {
 })
 
-xtest('extending styles', () => {
+test('extending styles', () => {
```
Did not spot that existing test. Thanks. Will remove

dumbNickname commented:
> I left a comment in the code, please let me know if it's clear.

Probably stupid me, but where can I find the mentioned comment? I do not see any new changes in origin repo. or this is something already in place? Please excuse bothering you again :P

MicheleBertoli commented:
Hey @dumbNickname, sorry for the misunderstanding I was referrring to [this comment](https://github.com/styled-components/jest-styled-components/pull/201#pullrequestreview-169115690) (it's not literally in the code, it's within the code using the GitHub review tool) - you already answered :)

I'm looking forward to receiving the update, and then merge.
Thank you very much!

dumbNickname commented:
@MicheleBertoli I removed superfluous tests. In case of nested component I saw that internal representation of styled component changed and now each components is wrapped with its own context so it looks like it will not work this way anymore. I also had the same problem in my own project and did not find a solution / nice alternative yet (in terms of testing)

I also included one more change  that I hope will not be controversial - it is related to difference in detecting lack of css value vs. css rule available but without a value. It is a quick fix, but I had no better idea. I added a test case to show the circumstances under which the old implementation was failing (btw. this was a case that I found in my project and worked well with previous jest-styled-components). Probably you will have a better idea were to put this test etc. If you do not agree with this change then also let me know,

MicheleBertoli commented:
Thank you very much for the update, @dumbNickname.

> I also included one more change

Would it be possible to submit a different PR for each change?
That would make things much easier.

dumbNickname commented:
@MicheleBertoli Sure it is possible ;) I removed the additional changes, rebased my branch on top of current origin and squashed some superfluous commits to have cleaner history. Please let me know if there is anything more I can do.

MicheleBertoli commented:
`7.0.0-2`
